### PR TITLE
Estimate Scheduler node runtimes

### PR DESCRIPTION
**Pull request by xmfan**
Working as starter task with @Chillee

This PR adds a method under BaseSchedulerNode to estimate the node's runtime in seconds. 

We use a heuristic based approach, first by considering whether the operation is memory bandwidth bounded or compute bounded:
- memory bandwidth bounded: we compute the number of bytes that are read/written to
- compute bounded: we compute the FLOPS required by the operation

One use case could be to be used as a cost model for scheduling: https://github.com/pytorch/pytorch/pull/100762

```
(pytorch-3.10) [14:08:02] ~/local/pytorch (xmfan/estimate_snode_runtime) > python3 test/inductor/test_perf.py -k EstimateSnodeRuntimeTests
[(ExternKernelSchedulerNode(name='buf0'), 400)]
[(ExternKernelSchedulerNode(name='buf0'), 2.35057908433887e-27)]
.[(ExternKernelSchedulerNode(name='buf0'), 3000), (SchedulerNode(name='buf1'), 3000)]
[(ExternKernelSchedulerNode(name='buf0'), 2.35057908433887e-26), (SchedulerNode(name='buf1'), 7.187055238190188e-09)]
.[(ExternKernelSchedulerNode(name='buf0'), 3000)]
[(ExternKernelSchedulerNode(name='buf0'), 2.35057908433887e-26)]
.[(ExternKernelSchedulerNode(name='buf0'), 34600)]
[(ExternKernelSchedulerNode(name='buf0'), 3.22687496698039e-24)]
.[(ExternKernelSchedulerNode(name='buf0'), 396)]
[(ExternKernelSchedulerNode(name='buf0'), 1.88046326747109e-27)]
.[(ExternKernelSchedulerNode(name='buf0'), 396)]
[(ExternKernelSchedulerNode(name='buf0'), 1.88046326747109e-27)]
.[(ExternKernelSchedulerNode(name='buf0'), 7776176)]
[(ExternKernelSchedulerNode(name='buf0'), 4.63240241413653e-21)]
.[(FusedSchedulerNode(nodes=buf0_buf1), 210)]
[(FusedSchedulerNode(nodes=buf0_buf1), 5.030938666733132e-10)]
.[(ExternKernelSchedulerNode(name='buf0'), 300)]
[(ExternKernelSchedulerNode(name='buf0'), 2.35057908433887e-27)]
.[(SchedulerNode(name='buf0'), 20)]
[(SchedulerNode(name='buf0'), 4.7913701587934585e-11)]
.
----------------------------------------------------------------------
Ran 10 tests in 14.311s
OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -360,6 +360,80 @@ def codegen_originating_info(self, buffer, only_once=True):
...
+
+        from .ir import CollectiveKernel, ComputedBuffer
+
+        # TODO(xmfan): figure out how to get hardware specs, use A100 for now
+        gpu_memory_bandwidth = 1555 * 2**30  # 1555 GBps
+        gpu_flops = 312**12  # 312 TFLOPS
+
+
+        def handle_extern_kernel(snode: ExternKernelSchedulerNode):
+            from torch.utils.flop_counter import (
```
Comment (Chillee) [inline torch/_inductor/scheduler.py:383]: Would be cleaner to do this using faketensors and flopcountermode instead of directly calling each operator separately.

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -360,6 +360,80 @@ def codegen_originating_info(self, buffer, only_once=True):
         buffer.writelines(out_lines)
         self.written = True
 
+    def get_estimated_runtime(self, rwbytes: int) -> float:
```
Comment (Chillee) [inline torch/_inductor/scheduler.py:363]: This probably shouldn't take in the rwbytes and the code to get the read/write bytes should be refactored to be in scheduler.py

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -360,6 +360,80 @@ def codegen_originating_info(self, buffer, only_once=True):
...
+
+        if "cuda" != layout.device.type:
+            # default to no reordering based on runtime
+            return 0
+
+        from .ir import CollectiveKernel, ComputedBuffer
+
+        # TODO(xmfan): figure out how to get hardware specs, use A100 for now
+        gpu_memory_bandwidth = 1555 * 2**30  # 1555 GBps
+        gpu_flops = 312**12  # 312 TFLOPS
```
Comment (eellison) [inline torch/_inductor/scheduler.py:379]: Reuse existing get_flops function (hardware agnostic), and factor it out into utils.py ? 

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L151-L165

That also shows memory bandwidth call `get_dram_gbps`

Comment (xmfan) [reply torch/_inductor/scheduler.py]: Addressed in 
a9bb8ea

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (xmfan) [reply torch/_inductor/scheduler.py]: Addressed in d1fb04c. `get_read_write_buffers_sizes` is computed twice, and it's a bit awkward to address. I'm calling `BaseSchedulerNode::get_estimated_runtime` from `GraphLowering::count_bytes` just to have a unit test cover it.

For prod, I'm expecting `get_estimated_runtime` to be called from elsewhere e.g. https://github.com/pytorch/pytorch/pull/100762, and those code paths will not need `get_read_write_buffers_sizes`, so we shouldn't have it as an output argument of `BaseSchedulerNode::get_estimated_runtime` just to avoid computing it a second time

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (xmfan) [reply torch/_inductor/scheduler.py]: Addressed in [ba0fdc2](https://github.com/pytorch/pytorch/pull/106426/commits/ba0fdc201a50fed3bbf66acbf20a1e6e507cb9f6)

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -362,6 +380,86 @@ def codegen_originating_info(self, buffer, only_once=True):
...
+
+        if "cuda" != layout.device.type:
+            # default to no reordering based on runtime
+            return 0
+
+        # TODO(xmfan): figure out how to get hardware specs
+        try:
+            from triton.testing import get_dram_gbps
+
+            gpu_memory_bandwidth = get_dram_gbps()
```
Comment (Chillee) [inline torch/_inductor/scheduler.py:430]: Move this into utils as well?

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -362,6 +380,86 @@ def codegen_originating_info(self, buffer, only_once=True):
...
+                    display=False
+                ) as flop_counter_mode:
+                    from .ir import ir_node_to_tensor
+
+                    fake_inputs = [
+                        ir_node_to_tensor(input) for input in self.node.inputs
+                    ]
+                    cls = self.node.__class__
+                    cls.process_kernel(op, *fake_inputs, **self.node.kwargs)
+                    return flop_counter_mode.get_total_flops() / gpu_flops
```
Comment (Chillee) [inline torch/_inductor/scheduler.py:453]: We might want a more nuanced strategy here. But to start with, might be a good idea to just multiply it by like ... 0.5

**Review by Chillee — APPROVED** [review]
LGTM! Some small nits

File: torch/_inductor/scheduler.py (hunk 4 above)
Comment (xmfan) [reply torch/_inductor/scheduler.py:430]: Not sure I understand, do you mean to alias triton.testing.get_dram_gbps as torch._inductor.utils.get_dram_gbps? or is this referring to the discussion for a compute device interface within pytorch?

I looked into triton's implementation, and it involves calling into hip/cuda C driver: driver.py: https://fburl.com/0vcm10ra, cuda.c: https://fburl.com/8cb6wi6j. There's a lot of code that would need to be copied

Comment (Chillee) [reply torch/_inductor/scheduler.py:430]: Yeah I mean to alias it into `torch._inductor.utils`, and ideally eventually abstract it into a compute device interface.

**Comment by Chillee** [comment]
By the way, one fun thing you might want to try is to run some models (like say, resnet18), and see how close your predicted runtime is to their actual runtime.
